hedera-node/src/main/java/com/hedera/services/store/contracts/precompile/TokenUpdateLogic.java

```diff
@@ -0,0 +1,246 @@
+import com.hedera.services.state.enums.TokenType;
+import com.hedera.services.state.merkle.MerkleToken;
+import com.hedera.services.store.contracts.WorldLedgers;
+import com.hedera.services.store.models.Id;
+import com.hedera.services.store.models.NftId;
+import com.hedera.services.store.tokens.HederaTokenStore;
+import com.hedera.services.store.tokens.annotations.AreTreasuryWildcardsEnabled;
+import com.hedera.services.txns.util.TokenUpdateValidator;
+import com.hedera.services.txns.validation.OptionValidator;
+import com.hederahashgraph.api.proto.java.AccountID;
+import com.hederahashgraph.api.proto.java.ResponseCodeEnum;
+import com.hederahashgraph.api.proto.java.TokenID;
+import com.hederahashgraph.api.proto.java.TokenUpdateTransactionBody;
+import com.hederahashgraph.api.proto.java.TransactionBody;
+import java.util.Optional;
+import javax.inject.Inject;
+
+public class TokenUpdateLogic {
+    private final OptionValidator validator;
+    private final HederaTokenStore tokenStore;
+    private final WorldLedgers worldLedgers;
+    private final SideEffectsTracker sideEffectsTracker;
+    private final SigImpactHistorian sigImpactHistorian;
+    boolean allowChangedTreasuryToOwnNfts;
+
+    @Inject
+    public TokenUpdateLogic(
+            final @AreTreasuryWildcardsEnabled boolean allowChangedTreasuryToOwnNfts,
+            OptionValidator validator,
+            HederaTokenStore tokenStore,
+            WorldLedgers worldLedgers,
+            SideEffectsTracker sideEffectsTracker,
+            SigImpactHistorian sigImpactHistorian) {
+        this.validator = validator;
+        this.tokenStore = tokenStore;
+        this.worldLedgers = worldLedgers;
+        this.sideEffectsTracker = sideEffectsTracker;
+        this.sigImpactHistorian = sigImpactHistorian;
+        this.allowChangedTreasuryToOwnNfts = allowChangedTreasuryToOwnNfts;
+    }
+
+    public void updateToken(TokenUpdateTransactionBody op, long now) {
+        final var tokenID = Id.fromGrpcToken(op.getToken()).asGrpcToken();
+        validateFalse(tokenID.equals(MISSING_TOKEN), INVALID_TOKEN_ID);
+        if (op.hasExpiry()) {
+            validateTrueOrRevert(validator.isValidExpiry(op.getExpiry()), INVALID_EXPIRATION_TIME);
+        }
+        MerkleToken token = tokenStore.get(tokenID);
+        checkTokenPreconditions(token, op);
+
+        ResponseCodeEnum outcome = autoRenewAttachmentCheck(op, token);
+        validateTrueOrRevert(outcome == OK, outcome);
+
+        Optional<AccountID> replacedTreasury = Optional.empty();
+        if (op.hasTreasury()) {
+            var newTreasury = op.getTreasury();
+            validateFalseOrRevert(isDetached(newTreasury), ACCOUNT_EXPIRED_AND_PENDING_REMOVAL);
+
+            if (!tokenStore.associationExists(newTreasury, tokenID)) {
+                outcome = tokenStore.autoAssociate(newTreasury, tokenID);
+                if (outcome != OK) {
+                    abortWith(outcome);
```

Comment:
Are these four aborts verified in E2E tests?  Either unit tests or E2E tests (or both) should trip these errors to ensure they fire.